icicle_v3/backend/cpu/include/ntt_template.h

```diff
@@ -0,0 +1,264 @@
+
+  template <typename S = scalar_t>
+  eIcicleError generate_twiddles(const S& primitive_root, S* twiddles, int n)
+  {
+    S omega = primitive_root;
+    twiddles[0] = S::one();
+    for (int i = 1; i <= n; i++) {
+      twiddles[i] = twiddles[i - 1] * omega;
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t>
+  eIcicleError cpu_ntt_init_domain(const Device& device, const S& primitive_root, const NTTInitDomainConfig& config)
+  {
+    // (1) check if need to refresh domain. TODO - reusing domain for different ntt sizes if possible
+    if (s_ntt_domain<S> != nullptr) { return eIcicleError::SUCCESS; }
+    // (2) build the domain
+    delete s_ntt_domain<S>;
+    s_ntt_domain<S> = new CpuNttDomain<S>();
+
+    bool found_logn = false;
+    S omega = primitive_root;
+    unsigned omegas_count = S::get_omegas_count();
+    for (int i = 0; i < omegas_count; i++) {
+      omega = S::sqr(omega);
+      if (!found_logn) {
+        ++s_ntt_domain<S>->max_log_size;
+        found_logn = omega == S::one();
+        if (found_logn) break;
+      }
+    }
+
+    s_ntt_domain<S>->max_size = (int)pow(2, s_ntt_domain<S>->max_log_size);
+    if (omega != S::one()) {
+      ICICLE_LOG_ERROR << "Primitive root provided to the InitDomain function is not in the subgroup";
+      return eIcicleError::INVALID_ARGUMENT;
+    }
+
+    // calculate twiddles
+    // Note: radix-2 INTT needs ONE in last element (in addition to first element), therefore have n+1 elements
+
+    s_ntt_domain<S>->twiddles = (S*)malloc((s_ntt_domain<S>->max_size + 1) * sizeof(S));
+    generate_twiddles(primitive_root, s_ntt_domain<S>->twiddles, s_ntt_domain<S>->max_size);
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t>
+  eIcicleError cpu_ntt_release_domain(const Device& device)
+  {
+    // release the memory
+    free(s_ntt_domain<S>->twiddles);
+    delete s_ntt_domain<S>;
+    s_ntt_domain<S> = nullptr;
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename E = scalar_t>
+  eIcicleError bit_reverse(int size, int logn, E* output, int batch_size)
+  {
+    ICICLE_LOG_DEBUG << "BIT REVERSE";
+    int total_size = size * batch_size;
+    for (int batch = 0; batch < batch_size; ++batch) {
+      E* current_output = output + batch * size;
+      for (int i = 0; i < size; ++i) {
+        int rev = 0;
+        for (int j = 0; j < logn; ++j) {
+          if (i & (1 << j)) { rev |= 1 << (logn - 1 - j); }
+        }
+        if (i < rev) { std::swap(current_output[i], current_output[rev]); }
+      }
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t, typename E = scalar_t>
+  eIcicleError cpu_ntt_ref(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output)
+  {
+    if (size & (size - 1)) { return eIcicleError::INVALID_ARGUMENT; }
+
+    int total_size = size * config.batch_size;
+    E* temp_elements = new E[total_size];
```

Comment:
prefer make_unique<S[]](size) over new.
It is automatically released so you don't leak even if you hit an exception or error that makes you return early.

Reply:
fixed